firedrake/interpolation.py

```diff
@@ -1664,3 +1666,94 @@ def multHermitian(self, source_vec, target_vec):
             # matrix will then have rows of zeros for those points.
             target_vec.zeroEntries()
             self.reduce(source_vec, target_vec)
+
+
+class ClementInterpolator(SameMeshInterpolator):
+    r"""
+    Compute the Clément interpolant of a :math:`\mathbb{P}0` source field, i.e., take
+    the volume average over neighbouring cells at each vertex.
+
+    See :cite:`Clement1975` for details.
+
+    For arguments, see :class:`.Interpolator`.
+    """
+
+    def __new__(cls, expr, V, **kwargs):
+        target_mesh = as_domain(V)
+        source_mesh = extract_unique_domain(expr) or target_mesh
+        if target_mesh is source_mesh:
+            return object.__new__(ClementInterpolator)
+        else:
+            raise ValueError("Clément interpolation requires the source and target meshes to coincide.")
+
+    @no_annotations
+    def __init__(
+        self,
+        expr,
+        V,
+        subset=None,
+        freeze_expr=False,
+        access=op2.WRITE,
+        bcs=None,
+        allow_missing_dofs=False,
+    ):
+        if subset:
+            raise NotImplementedError("subset not implemented")
+        if freeze_expr:
+            raise NotImplementedError("freeze_expr not implemented")
+        if access != op2.WRITE:
+            raise NotImplementedError("access other than op2.WRITE not implemented")
+        if bcs:
+            raise NotImplementedError("bcs not implemented")
```

Comment:
This sort of thing makes me wonder whether this is the best place to implement this functionality. Might it be better to simply have a free function called `interpolate_clement` that does the parloop?

Reply:
As in instead of having the `Interpolator` at all?

Reply:
Yes. I think we need @dham to weigh in here. Is this the right abstraction?

Reply:
It might be nice to have `assemble(interpolate(source, target, variant="clement"))` but that might be tricky/unwieldy.

Reply:
I think we'd have to think harder about whether that is actually the Right Thing. For a finite element space with a known dual basis, their is a single canonical interpolation operation. Other interpolation operations are meaningfully different: they are just other projections that don't directly arise from the definition of the space.

Reply:
The variant option at least requires heavier surgery because `interpolate` is defined in UFL.